src/Simulation/Native/src/external/fusion.hpp

```diff
@@ -58,7 +64,48 @@ class Fusion{
 		handle_controls(empty_matrix, empty_vec, {}); // remove all current control qubits (this is a GLOBAL factor)
 	}
 	
+	IndexSet getSet() const {
```

Comment:
You bring out a very interesting point here, and I noticed that you brought this up in multiple places. However, making it a reference takes us away from the getFoo/setFoo design pattern used throughtout the repository. This change would require a significant overhaul in many aspects of the codebase. We could look into making use of std:move, however it would be appropriate to get back to this as part of a another PR linking it to an appropriate issue. Given that the data structures these work on are set of gates, I believe copying does not add much overhead. right now. Thank you for pointing this out!

Reply:
If a design pattern leads to a lesser quality code, the design pattern shouldn't be used. You can keep the name as get_foo, if you so prefer (though I personally like to reflect the semantics in the name), the callsites then won't have to change but won't incure the copy costs _unless they need their own copy of the set_, which they can still do explicitly.

The getters you are adding are wasteful and can be easily fixed without the overhead of a follow-up PR that would have to cleanup the mess.

Reply:
Switching to a reference pattern would also mean removing the setters, wouldn't it? Each time get is called, any changes made to the returned structure would already be on the internal structure. I wonder if a move-based approach would work better, which would have the added benefit of only letting one part of the code own the data structure at a time (kind of like the Rust semantics).

That said, given how small these data structures are, this is a minor enough optimization that I think noting it and addressing it in a later change would make sense.

Reply:
returning const& prevents modifying the vectors by the clients and is usually the most efficient way to allow read-only access to the data. The setters that swap in a new data-structure as a whole are still valuable and have different semantics compared to allowing modifying the data in place. The structures aren't that small either (vectors of vectors plus other stuff) in some of the cases) -- why return by value if there is an established C++ pattern for such scenarios?

Reply:
I have added const&, and it seemed that the callsites did not required any change as I anyway was just using them to create new structures which eventually get set later